### ISSUE TITLE
Modeling disaggregate "cost" in VF

### ISSUE THREAD

**Issue by orthecreedence** [valueflows/valueflows#616]
I am embarking to convert an existing project ([Basis](https://basisproject.gitlab.io/public/)) to use VF. One of the core concepts of Basis is that it tracks costs of products (a set of costs being a collection of both resources like raw materials and labor hours/wages of various occupations). These costs are not really "exchanged" between companies in the traditional sense, where ownership is transferred in exchange for money, but rather the costs are imbued in the products themselves and once transferred, are added into various cost buckets within the receiving company.

I think I have a grasp on how a traditional transaction might be modeled (ChipFab transfers processors to PhoneCo, PhoneCo transfers $10000 to ChipFab, modeled as two `EconomicEvent`s). In the case where ChipFab transfers processors to PhoneCo and PhoneCo *doesn't* transfer anything back but rather adds the costs of the chips (ex `{"labor": {"val": 500, unit: "hours"}, "silicon": {"val": 1000, "unit": "grams"}, ...}`) to their inventory costs, I'm not sure how this would look.

I'd love some insight on the best way to model something like this. Thanks!

**Comment by bhaugen** [comment]
> I am embarking to convert an existing project ([Basis](https://basisproject.gitlab.io/public/)) to use VF. 

That should be very interesting! We're happy to help.

> One of the core concepts of Basis is that it tracks costs of products (a set of costs being a collection of both resources like raw materials and labor hours/wages of various occupations). These costs are not really "exchanged" between companies in the traditional sense, where ownership is transferred in exchange for money, but rather the costs are imbued in the products themselves and once transferred, are added into various cost buckets within the receiving company.

I've been intrigued by the Basis approach to product costing and transactions and would like to follow all the implications, which I have not done yet. Might need a real-life experiment...

> I think I have a grasp on how a traditional transaction might be modeled (ChipFab transfers processors to PhoneCo, PhoneCo transfers $10000 to ChipFab, modeled as two `EconomicEvent`s). In the case where ChipFab transfers processors to PhoneCo and PhoneCo _doesn't_ transfer anything back but rather adds the costs of the chips (ex `{"labor": {"val": 500, unit: "hours"}, "silicon": {"val": 1000, "unit": "grams"}, ...}`) to their inventory costs, I'm not sure how this would look.
> 
> I'd love some insight on the best way to model something like this. Thanks!

ValueFlows does not require that resource transfers elicit reciprocal resource transfers.

I'll describe the PhoneCo and ChipFab transaction first from the model perspective, and then from the software implementation perspective.

From the model perspective, the chips were manufactured in a Process where inputs of labor, silicon, etc, were recorded. The silicon came from some other processes with other inputs. The labor came from a person who had a bunch of needs.

So you can traverse the resource flows back as far as it makes sense, or as far as you can access the data. You can read more about that topic in https://github.com/valueflows/vf-apps-traversing-the-flows

Here's a screenshot from a project that was a precursor to ValueFlows and uses pretty much the same model: https://speakerdeck.com/mikorizal/10-nrp-value-equation-concepts-and-tutorial?slide=16

From a software implementation perspective, you could either traverse the flows on-the-fly whenever you want the information, or aggregate it in some data structure connected to the various resources when they are created. I expect that Basis may already have such structures.

@orthecreedence did that make sense?

There's [pointers to Python code for doing the traversals in this issue](https://github.com/valueflows/django-vocabulator/issues/7), and Rust code will be forthcoming from the Holo-REA project.

**Comment by orthecreedence** [comment]
> That should be very interesting! We're happy to help.

Thanks!

> I've been intrigued by the Basis approach to product costing and transactions and would like to follow all the implications, which I have not done yet. Might need a real-life experiment...

I need to work on messaging/visuals a bit more, probably. This is something I haven't spent much time figuring out. How did you create the visuals for VF?

> So you can traverse the resource flows back as far as it makes sense, or as far as you can access the data. You can read more about that topic in https://github.com/valueflows/vf-apps-traversing-the-flows

Ah, this is the piece I was missing! In Basis, I originally started out with the idea of graph traversal, but ultimately ended up wanting to record the costs embedded in the products for two reasons:

1. Trust. If the system processing the transactions and tallying the costs is also the system tracking the end-costs, then there's no need for a second process that reads the data and calculates the costs after the fact, which might be subject to its own amount of tampering (for instance a mobile app that fudges the numbers based on who the donors are). Having the system embed costs as they accumulate/divide makes knowing "what does this cost?" a simple query vs a fairly in-depth process. It might be possible to build traversal into the same system that tracks the transactions, and maybe this is something holochain would be better at, but doing something like this in blockchain would grind the transaction processing to a halt.
2. Performance. Graph traversal with a few hops is easy, but once we get into recursive loops and extremely long supply chain graphs, the question "what does this cost" becomes much harder to answer, or even prohibitive depending on how many loops there are in the economy. It becomes less about counting the hops and more about detecting loops and building in some sort of decay such that infinite loops don't happen. Embedding the costs as they move through the system encapsulates them at the moment of creation such that they forego unneeded processing. This also handles endless economic loops "naturally" because all that's really required is knowing a) the inputs over time b) the outputs over time and c) the ratio of how inputs match to the outputs (done currently via "buckets"). Having each "node" in the graph calculate this on its own completely eliminates the need for graph traversal and trades the processing power needed for traversal for what amounts to storage.

> Here's a screenshot from a project that was a precursor to ValueFlows and uses pretty much the same model: https://speakerdeck.com/mikorizal/10-nrp-value-equation-concepts-and-tutorial?slide=16

Thanks, this helps bolster my understanding of how this would work. While I still might avoid graph traversal, knowing how to do it will ultimately allow me to verify the results of tracked aggregate costs more accurately. This slide/flow is really useful in general.

> From a software implementation perspective, you could either traverse the flows on-the-fly whenever you want the information, or aggregate it in some data structure connected to the various resources when they are created. I expect that Basis may already have such structures.

Ok, this makes sense! I should have read your entire comment before going on my numbered rant above (sorry). So the idea would be to attach the costs as a separate data structure. I'm fine with this, especially if it's your recommendation. My fear was that I would do something like this and a few weeks later someone would say "why didn't you just use `vf:resourceCostVector`??" or something ;-).

I'm looking forward to learning more and modeling this out. Thanks for your help.

**Comment by orthecreedence** [comment]
> We tend to be more minimalist when it comes to the model and more pragmatic when it comes to getting stuff done.

I like this approach.

> Like if you want an instant answer, go with the pre-aggregated data. If you want to be sensitive to changes in the historical data, you might need to have signals of change bubble up from the preceding flows and re-aggregate. If you want to explore all of the details of provenance, you may need to traverse the flows.

This is true, changes to historical data are not something I considered fully. For instance I would count a "return" as a new transaction/cost rather than a modification, but hadn't considered some sort of either nullification or anything of that nature. Good things to think about.

Traversal might make sense when thinking about holochain as a medium for transactions because one node doing a traversal doesn't block other nodes from processing (whereas a blockchain transaction would). I'll keep this in mind when evaluating holochain.

Also, Basis is more interested in cost over time (so really, cost of `ResourceSpecification` vs cost of `EconomicResource`) so modification in the history of the chain would likely have a minimal effect on cost when averaged over the period of some amount of time (a year or five years for example). This might make things more complicated when traversing because you'd potentially have to look through the data over a long period of time for each company.

However, maybe the aggregation could happen on a lower level...like, maybe the aggregate cost-over-time is stored in each `Agen:ResourceSpec` pair (as opposed to storing the cost collection with each order itself) and then tree traversal could be used to find the costs based on those aggregates for each of the products. Sort of a hybrid approach. 

> But `vf:resourceCostVector` might be a good idea....???

It's probably overkill for now, especially because even in my own head the idea is in flux. After [this Reddit conversation](https://www.reddit.com/r/basisproject/comments/f70ly4/heres_a_pointer_to_a_possibly_related_project_ive/fizs6fr/) I think it makes sense to change the cost storage to be a tree (or array of trees) where the leaf nodes are the current "costs" and the connection between the nodes and their respective parents would represent a `vf:Process` that transformed some tracked resource into another tracked resource.

In other words, I wouldn't want to push for you to change your model until I had a solid idea of what mine looks like. 

Thanks, this discussion is giving me some interesting things to think about...

**Comment by bhaugen** [comment]
> I wouldn't want to push for you to change your model until I had a solid idea of what mine looks like.

Aggregating inputs in the incoming resource flows into some kind of data structure would be an implementation thing. Wouldn't change the model, which would still have graphs and trees of input-process-output connections.

But if you come up with some interesting implementations, always nice to share.
